extensions/amp-lightbox/1.0/amp-lightbox.js

```diff
@@ -50,31 +50,36 @@ class AmpLightbox extends PreactBaseElement {
     this.registerApiAction('open', (api) => api.open(), ActionTrust.LOW);
     this.registerApiAction('close', (api) => api.close(), ActionTrust.LOW);
     return dict({
-      'onBeforeOpen': this.beforeOpen_.bind(this),
-      'onAfterClose': this.afterClose_.bind(this),
+      'onBeforeOpen': this.toggle_.bind(this, true),
+      'onAfterClose': this.toggle_.bind(this, false),
     });
   }
 
-  /**
-   * Setting hidden to false
-   * @private
-   */
-  beforeOpen_() {
-    this.open_ = true;
-    toggle(this.element, true);
-    this.element.setAttribute('open', '');
-    fireEvent(this.win, this.element, 'open', ActionTrust.HIGH);
+  /** @override */
+  triggerEvent(element, eventName, detail) {
+    const event = createCustomEvent(
+      toWin(element.ownerDocument.defaultView),
+      `amp-lightbox.${eventName}`,
+      detail
+    );
+    Services.actionServiceForDoc(element).trigger(
+      element,
+      eventName,
+      event,
+      ActionTrust.HIGH
+    );
+
+    super.triggerEvent(element, eventName, detail);
```

Comment:
Two points:
(1) Can `ActionService.trigger` start firing `triggerEvent`/`dispatchCustomEvent` inside? Fine not to do it in this PR, but might be a better layer for that.
(2) I think there should also be a `custom-element.js` that just calls `triggerEvent`.

Reply:
1. Widening the question a bit: components seem to fire an `amp-my-component.fooBar` custom event through the action service in addition to a no-namespace `fooBar` custom event (via `dispatchCustomEvent`). Currently, I have `PreactBaseElement#triggerEvent` dispatching "plain" events, while AMP components sometimes override to also trigger a similar event through the action service. Reason for this: Bento code can call `this.triggerEvent(...)` as appropriate so more core logic can live on Bento side, AMP component can hook in and fire the AMP-side events

If I'm understanding correctly, your suggestions lead to three AIs:
- `ActionInvoctation#trigger(el, namespacedEventName, evt, trust)` should perform its normal functions then `dispatchCustomEvent(el, plainEventName, ...)`
- define `CustomElement#trigger` to trigger through action service (which would also handle dispatching the "plain" custom event
- AMP elements can override `triggerEvent` to trigger through action service like above, without the need to call super anymore if action service is handling dispatching both events

Does that sound right to you? If so I'll file a tracking issue; if not, could you let me know where I'm misunderstanding?

Reply:
2. I think `CustomElement#dispatchCustomEvent` was removed (https://github.com/ampproject/amphtml/pull/31058) as a public API on the elements themselves, partially because the event was getting triggered on not-the-best element rather than letting calling code explicitly provide the best one. So, maybe we should not re-add `CustomElement#trigger`

Reply:
Thinking about 1 but wanted to quickly respond to 2: I meant inside `amp-lightbox/custom-element.js`, not `src/custom-element`

Reply:
I think I've lost track of the thread here; there isn't an `amp-lightbox/custom-element.js` file (there's `component.js`, `base-element.js`, and `amp-lightbox.js`). Currently, `component.js` receives the `onBeforeOpen`/`onAfterClose` as props from `AmpLightbox#init`, and it calls those, which in turn call both `toggle` and `triggerEvent`. Are you suggesting moving the `toggle` logic entirely within `LightboxWithRef` in `component.js`?

Reply:
You're right, I was recommending creating that file, but I think this suggestion is maybe working off stale knowledge of the suggested file structure and each file's purpose. Maybe it goes in `base-element.js` instead? The important point is that the DOM events should fire in non-AMP HTML documents as well, and the AMP events should fire in AMP documents only.

Reply:
re: 1 - 

> - ActionInvoctation#trigger(el, namespacedEventName, evt, trust) should perform its normal functions then dispatchCustomEvent(el, plainEventName, ...)

Yes, that's correct. In valid AMP documents this should have no impact. Though I guess if we do not care to support the case where invalid AMP documents with the runtime also contain custom JS, maybe this is not as crucial to do after all. I'm trying to think if there are other cases someone would benefit from having both AMP Action system and DOM events to interact with.

> - define CustomElement#trigger to trigger through action service (which would also handle dispatching the "plain" custom event
> - AMP elements can override triggerEvent to trigger through action service like above, without the need to call super anymore if action service is handling dispatching both events

I was not recommending either of these steps, and I believe this was a result of my misunderstanding the concept of an `amp-lightbox/custom-element.js` that may not be relevant here. I would consider these to be resolved with our discussions regarding (2) though.

Reply:
1. Filed tracking issue https://github.com/ampproject/amphtml/issues/32756 , will put together draft change during Fixit 
2. Got it :+1: